### ISSUE TITLE
Blender Content in LibGDX Code

### ISSUE THREAD

**Issue by MatthewGitter** [GoranM/bdx#839]
Hi. Where is the content generated in Blender (meshes, materials, textures, properties) stored in the LibGDX project? I see the sacky.png texture in android/assets/bdx/textures, but I don't see any sign of the meshes, properties or other content that defines the Blender objects, even though I "Make internal java files external" and I see Sacky.java etc in my core module. Is it in android/assets/bdx/scenes/Scene.bdx that Scene.class in bdx.jar has instructions to render?

**Comment by GoranM** [comment]
Yes, it's in the .bdx files, which are generated for each scene.

**Comment by MatthewGitter** [comment]
So all meshes are serialized into ASCII in a single Scene.bdx file? I added the Stanford bunny.ply to the scene, and that 15KB .PLY file added 6.5MB to Scene.bdx, though the .blend file is only 615KB. That's not going to scale.

**Comment by rafcolson** [comment]
>So all meshes are serialized into ASCII in a single Scene.bdx file?

Not sure what you mean by that. If you open `Scene.bdx` in a text editor you'll see that it's a (json) dictionary. It's generated in `exported.py` [here](https://github.com/GoranM/bdx/blob/master/blender/bdx/exporter.py#L868).

**Comment by MatthewGitter** [comment]
I looked. The meshes are serialized in JSON, which is ASCII, which is over 400x the bytes of even an ASCII PLY file containing the same bunny.ply mesh. The JSON is over 10x the bytes of the .blend file containing the bunny mesh after it's imported into Blender. 100 models like the bunny would make a 650MB Scene.bdx, even before logic and other assets (textures) which could easily make it 1GB (or across multiple Scene.bdx files, which are probably larger due to overhead per scene file.

Is the Scene.bdx file downloaded by the HTML module when it's run? Is the Scene.bdx module bundled into the Android APK? If so, that's not going to be a viable download/install at those scales.

**Comment by SolarLune** [comment]
> I added the Stanford bunny.ply to the scene, and that 15KB .PLY file

Which bunny is this? The ones I got from the official (?) website are larger than that in size, even in the lowest resolution. How many verts and tris does it have, as well? Are you using the highest resolution one - the one that has ~1M tris?

> 100 models like the bunny would make a 650MB Scene.bdx, even before logic and other assets (textures) which could easily make it 1GB (or across multiple Scene.bdx files, which are probably larger due to overhead per scene file.

Besides the bunny (because I don't know much about your particular bunny), logic and textures aren't included in the .bdx file. Logic is exported to `.java` files and put in the `core/src` folder if you type it up in Blender - generally you'd use an editor and put it there in the first place, though. Textures aren't included in the `.bdx` file either - they're referenced from their texture directory.

> (or across multiple Scene.bdx files, which are probably larger due to overhead per scene file.

I'm not quite sure what you mean by this.

> Is the Scene.bdx file downloaded by the HTML module when it's run? Is the Scene.bdx module bundled into the Android APK? If so, that's not going to be a viable download/install at those scales.

Just to make sure you're aware, each scene in Blender gets exported to a `.bdx` file, not just "one". So it's not `Scene.bdx` per se, it's different `.bdx` files for each scene  in Blender (i.e. `Level1.bdx`, `Level2.bdx`, `GUI.bdx`, etc). I'd assume the scene files are downloaded to fully load the game scene - GWT handles that. Same thing for the Android APK - it's got to be stored somewhere.

**Comment by MatthewGitter** [comment]
Well, I was mistaken about the bunny.ply file size, which is actually 658744B , not 15KB, at 8171 vertices and 16301 faces. But the point is that it becomes 10x as large in the Scene.bdx file. If I have 100 models that size, I've got a 650MB Scene.bdx file. And if I have those 100 models spread across different .bdx scene files, like 3 models in Level1.bdx, another 3 in Level2.bdx etc for 33 files, that's a larger total size of those 33 files than even a single Scene.bdx would be with 100 files. That's pretty big.

AFAICT the game.blend file also contains the same model data, but it's stored in a way that the bunny data added only 526142B to a 88757B game.blend file, which is even smaller than the bunny.ply (ASCII) file. Since that file format stores the same meshes (and probably other objects) in 1/10th the size the .bdx format does, it's a lot more workable for downloading a good amount of models every time the user hits the HTML version or for downloading and storing it on an Android's relatively small file system when installing the app (Iphone too).

**Comment by GoranM** [comment]
We're aware of these inefficiencies, and there was even some interest to address them in #538, but that seems to be on hold for now.

Since BDX was intended to support independent developers who aim to make interesting, but resource-light games, efficiency (and especially memory efficiency) was never a big priority.

**Comment by MatthewGitter** [comment]
It looks like aliasaila blocked on #538 because libGDX was stuck on GWT v2.6.0 with a bug in StringBuilder.toString() . But as of 11/16/16 their team was [about to upgrade GWT](https://github.com/libgdx/libgdx/issues/4389#issuecomment-260925597) to v2.8.0 (for Java 8 etc). The bug was fixed in GWT v2.7.0 . So maybe that hold can be removed.

I'm interested in using BDX to add code to large poly count models, possibly millions of vertices. If I can have Android and HTML versions with hundreds of megabytes instead of gigabytes that's important. Also if the models are in files external to the Scene.bdx JSONs, they could perhaps be downloaded on demand instead of monolithically with the entire scene.

**Comment by GoranM** [comment]
I don't know why the effort in #538 stalled. It could be the GWT issue, or a number of other things.

As I said before, it's not really a high priority, because BDX was designed for developers who can make interesting games without requiring "millions of vertices".

Actually, in a way, I think the inefficiencies are somewhat beneficial, because they encourage developers to be a little more creative, instead of just trying to plaster a scene with heavy assets.

**Comment by MatthewGitter** [comment]
I'm working with Google Tango, the AR capture 3D camera. It will capture individual models with hundreds of thousands or millions of vertices. I'd like to use BDX to add behavior to the models, for cross platform presentation, while editing them in Blender. These AR use cases will become common as more 3D model capture cameras come on the mass market starting with the January 2017 CES.

I'd be interested in coding factoring the serialized out of the JSON into a separate, smaller file, if that kind of feature would be welcome in BDX.

**Comment by SolarLune** [comment]
> I'd be interested in coding the factoring of the serialized model out of the JSON into a separate, smaller file, if that kind of feature would be welcome in BDX.

That seems to basically be the work done in #538, as @aliasaila prototyped. [The code's still there](https://github.com/GoranM/bdx/issues/538#issuecomment-185060387), so it shouldn't be too hard to tweak to work under BDX proper. Please feel free to look into this - I know I'd really appreciate smaller, quicker-loading, and more clearly-defined model files. My current project's not huge by any stretch, but it still takes a few seconds longer than I'd like to start-up; it might be due to model export. I'd have to test to make sure, though, of course.

Note that lots and lots of faces won't work under BDX for a second reason, which is that there can only be so many vertices on a single mesh (~32k, as I recall). This is a limit that is imposed by the length of the float array used to create the meshes, if I recall? I believe this is partially because of the data type used - an unsigned short might double the number of vertices (but it'd still be limited). [See the mesh creation process here.](https://github.com/GoranM/bdx/blob/master/src/com/nilunder/bdx/Scene.java#L731) 

[Interestingly, Unity (still?) seems to have the same issue.](http://answers.unity3d.com/questions/255405/vertex-limit.html)

Vertices are also duplicated across adjacent faces at the moment, which might be able to be resolved to more efficiently use the vertices within the limit. 

Lastly, automatic mesh splitting would probably be the best (?) solution to ensure large models load in without incident regardless of the internal vertex limit, though that would probably need to be made with the material slot system in mind. It's something else to consider.

> Also if the models are in files external to the Scene.bdx JSONs, they could perhaps be downloaded on demand instead of monolithically with the entire scene.

Hmm. The only idea I'd have for this is some method to tweak where the model path is for the project in Blender so that it could be external to the finished application instead of bundled within. That bridge could be crossed later, though; for now, just investigating the code @aliasaila wrote might be a solid path forward.

**Comment by MatthewGitter** [comment]
I'll take a look at the existing code.

Yes, the libGDX OpenGL calls use a short for the vertices buffer index, which since Java makes a signed short on a non-negative value means any mesh must max at 32K vertices. But there are MeshPart objects for splitting meshes. Mesh reduction would be a later optimization if ever. Dynamically downloaded meshes also later.

**Comment by GoranM** [comment]
Thinking about it further: I'm not really sure that aliasaila's approach is the best way to address these issues, and I would probably say the same for any approach that requires more export code, and more files.

It would be far better to just read the data directly from the .blend, bypassing the whole export step, and simplifying the system as a whole.

I wanted to do that when BDX was initially designed, but I couldn't find good documentation that describes the blender file format.

If someone would be willing to do that research now, and to gather up the relevant info, that would be a step in the right direction.

**Comment by SolarLune** [comment]
I remember reading about a Java library made to read .blend files - [this might be the one I recall.](http://homac.cakelab.org/projects/JavaBlend/spec.html#BlenderDNA) I see the benefit of reading from .blend files directly, as it seems simpler and possibly faster overall (especially without any export code). However:

1) Looks like the source code's available on the download page (alongside the binary). Also looks like the source is GPLv3 licensed, but the binaries are Apache v2. Something to keep in mind if we ever need to modify the source if the creator can't keep up with Blender as it's updated.

2) It doesn't really make sense to read blend files for just mesh importing when we also export other scene-specific data. That means that we would probably end up refactoring BDX as a whole to read from .blend files (basically wherever JSON is involved). Considering everything's pretty centrally located, though, I don't feel like it'd be _too_ difficult. I'm not sure if the project linked would cover all of the necessary cases (like game properties), though.

3) Security now becomes a bit more of a concern; `.jar` files are basically just ZIPs as far as I know, so with all assets and their placements in a scene being openly available as a `.blend` might put some people off. There wouldn't really be much stopping people from having almost all of your assets easily at hand for usage. It's something to consider, anyway.

**Comment by GoranM** [comment]
Refactoring BDX to read everything we need, directly from the .blend, was what I meant.

The library you referenced seems a little too complicated, and not just because there's so much code, but also because it seems to require a code generation step at build time.

It should be possible to write a fairly small piece of pure Java code that can read the relevant data from the .blend at runtime.

As for security: One could write a mangler to "scramble" the shipping .blend into some unrecognizable binary blob, and then have code that could reconstitute the original file in memory, which BDX can then load normally. That would make it about as difficult as extracting the data from the .bdx files.

... Although, I'm not sure if doing that is worthwhile, because I have yet to hear of a case where the original developer was somehow damaged by making their game data accessible to commonfolk. Actually, that level of access has historically resulted in higher popularity for both the developer and the game in question.

**Comment by SolarLune** [comment]
> It should be possible to write a fairly small piece of pure Java code that can read the relevant data from the .blend at runtime.

Not sure exactly what you mean - blend files are binary files, as I recall? It's not like you can "read" a blend file's text like you can a JSON file or an XML or something. 

How would we write a small piece of Java code that can read in meshes, let alone all of the stuff we need, without using a separate library?

> ... Although, I'm not sure if doing that is worthwhile, 

Yeah, I mean, `.jars` are pretty "open" already, so I don't think there's a huge reason to go for added security by using a mangler. It could be something done later if necessary, though.

**Comment by MatthewGitter** [comment]
libGDX configs to work with Proguard and Dexguard obfuscators:
https://github.com/libgdx/libgdx/wiki/ProGuard-DexGuard-and-libGDX

**Comment by MatthewGitter** [comment]
Sorry, clicked wrong comment button.

**Comment by GoranM** [comment]
Fundamentally, reading data from a binary file doesn't require a "separate library".

It's a matter of understanding the file format, so that the relevant bytes can be located, and then converted into the appropriate Java type.

The conversion is typically done with a ByteBuffer, but I'm not sure if that's supported by GWT, so we may have to use more primitive methods, as shown here: http://stackoverflow.com/questions/4513498/java-bytes-to-floats-ints/4513578#4513578

> Yeah, I mean, .jars are pretty "open" already

Even if they weren't, I think the developer could still benefit from making an opening of sorts, because, historically, that kind of access has proven to be a net positive for everyone involved.

**Comment by MatthewGitter** [comment]
The Blender project maintains [the bpy.data Python module](https://www.blender.org/api/blender_python_api_current/bpy.data.html) for parsing .blend files.

Blender uses Python 3.x. If Jython worked with Python 3.x it could compile bpy into Java bytecode to be integrated with BDX, but [Jython for Python 3.x](https://wiki.python.org/jython/JythonFaq/GeneralInfo#What_is_the_current_status_of_Jython.3F) is not complete.

**Comment by SolarLune** [comment]
> Fundamentally, reading data from a binary file doesn't require a "separate library".

> It's a matter of understanding the file format, so that the relevant bytes can be located, and then converted into the appropriate Java type.

I get the concept, it's just that I can't imagine doing this in "a fairly small piece of pure Java code". The library I linked shows a bit about the `.blend` file format - it doesn't seem particularly simple or straightforward.

The point of reading `.blend` files directly is to be quicker and simpler than exporting data manually, right? It seems that simplicity is lost if we have to write an importer (and a byte-based one, at that) anyway.

> but also because it seems to require a code generation step at build time.

What do you mean?

**Comment by GoranM** [comment]
I think it can be fairly small relative to the java .Blend library, because we're just concerned with getting the data that we need, instead of all the other features that java .Blend aims to provide.

It would be a chunk of code, but I don't think it would be more complicated than the code currently involved in the export process. Also, getting rid of all the BDX generated files, and just dealing with the .blend would be a significant system simplification.

> What do you mean?

On the page you linked, in the first paragraph:

> Java .Blend aims to fill the gap for Java programmers, who want type-safe access to all the data in a Blender file with read and write capabilities. To achive that, Java .Blend uses meta data which is contained in each Blender file, and describes Blender's entire core data model. This meta data is fed to Java .Blend's code generator, which generates Java classes for all types in Blender's data model and thereby creates a type-safe API to access raw data in a .blend file.

---

> The Blender project maintains the bpy.data Python module for parsing .blend files.

As far as I know, that module simply provides an api to access data in the running instance of Blender. It doesn't parse .blend files, or provide any utilities to do so.

However, even if it did:

> Blender uses Python 3.x. If Jython worked with Python 3.x it could compile bpy into Java bytecode to be integrated with BDX, but Jython for Python 3.x is not complete.

Since the proper operation of the bpy module depends on the running blender environment, it can't be made to work on the JVM by simply compiling its python code into java bytecode. And even if that could be done (somehow), GWT can only compile Java source code into JavaScript, so it wouldn't be able to run the raw java bytecode in the browser.

**Comment by MatthewGitter** [comment]
How about refactoring the [Cakelab JavaBlend SDK source](http://homac.cakelab.org/projects/JavaBlend/downloads/JavaBlend-SDK-src-1.0.4.zip) package org.cakelab.blender.io classes. Make them just read mesh data from the .blend file into created BDX mesh objects, instead of JavaBlend's own more complex approach of generating Java class sourcecode from the meshes. The refactored IO package would execute at runtime (in the libGDX executable modules).

Maybe later add a gradle task that compiles a little executable .jar (part of the regular build process) wrapped around the IO package then executes it (all at coding time) to generate a Scene.bdx file instead of the .blend file, for "classic BDX" behavior. A good complement to the existing "Make internal java files external" and "Export and run" functions.

**Comment by GoranM** [comment]
I'm not sure how easy it would be to refactor it in that way. There's also the question of whether that refactored code would be less complex than something we could write specifically for our needs.

**Comment by SolarLune** [comment]
So someone over on BlenderArtists was looking into making a C++ blend file importer, and someone posted some resources about the `blend` file format. Here's a [page](http://www.atmind.nl/blender/mystery_ot_blend.html) about the format, and here's how [Assimp](https://github.com/assimp/assimp/blob/master/code/BlenderLoader.cpp) loads 'em.

**Comment by SolarLune** [comment]
So I would really like to see a more improved workflow, as #853 indicated, so let's discuss this with that in mind. Reading blend files directly would be a very nice, efficient idea. Here's the problems currently:

1) I personally think it'd be a lot of work to write a binary blend file reader, but I've never done it before, so I don't really know. Barring using a library to read it for us or an example that gives the basic process and a starting point, I'm probably not going to investigate into this further. If anyone else is interested or has ideas, please post. One alternative might be to create a program to export data from Blender to an intermediary format using CakeLab's library (or another library, or just using another `.blend` reading library that doesn't generate code). It could be that that is the best of both worlds - faster than the current Python-based export process, and fairly simple to write.

2) Even with direct blend file reading, we would still have to write something to import mesh data, armatures, and animations, and bind animations to meshes. It might be more efficient, in terms of usage of time, to export the meshes to an intermediary format, like `.fbx`, and use LibGDX's normal importer to proceed. The up-side of this approach is that we should be able to get armatures and animation "for free", in a sense. The `.fbx` format is also a binary format, so for mesh data, it would load faster and compress smaller than the current JSON `.bdx` format. We would also use the AssetManager to load the models, armatures, and animations, so asynchronous loading should also come fairly easily.

Basically, the problem (in addition to the ones listed in #853) is:

`.bdx` doesn't compress well and takes a long time to import and export, mainly because of mesh data. How can we improve this?

**Comment by MatthewGitter** [comment]
I have been discussing the [Java.Blend project](http://homac.cakelab.org/projects/JavaBlend/) with its author. J.B incorporates the entire Blend DNA file format, reading all the .blend metadata and generating Java classes that represent the entire contents of the file. BDX could run J.B when it exports the LibGDX project and the .blend file from Blender, including the classes from the .blend file. This would be much like BDX's "Make internal java files external" in Blender. The J.B generator could be run again from the LibGDX project as a Gradle task.

The author, homac, just generated a newly refactored example project incorporating J.B into a Blender viewer that I can probably share with you if you're interested. This approach seems perfectly complementary to BDX.

**Comment by GoranM** [comment]
BDX generates certain java files because it has to do so, not because it's a good idea. Having a generator, and a bunch of generated files, is more complex, and opens up a lot of different vectors for potential problems.

Also, from what I understand, the code generated by Java.Blend uses native methods that are not GWT compatible, so even if we were to ignore the overhead of yet another generation step, it would still be something that we can't really use.
